presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -1114,35 +1117,51 @@ public PlanWithProperties visitUnion(UnionNode node, Context context)
                 }
             }
 
-            PlanNode result = null;
-            if (!partitionedChildren.isEmpty()) {
-                // add an exchange above partitioned inputs and fold it into the
-                // set of unpartitioned inputs
-                // NOTE: this must provide the explicit imput mapping as unions may drop or duplicate symbols
+            PlanNode result;
+            if (!partitionedChildren.isEmpty() && unpartitionedChildren.isEmpty()) {
+                // add an gathering exchange above partitioned inputs
                 result = new ExchangeNode(
                         idAllocator.getNextId(),
                         GATHER,
                         REMOTE,
                         new PartitioningScheme(Partitioning.create(SINGLE_DISTRIBUTION, ImmutableList.of()), node.getOutputSymbols()),
                         partitionedChildren,
                         partitionedOutputLayouts);
-
-                unpartitionedChildren.add(result);
-                unpartitionedOutputLayouts.add(result.getOutputSymbols());
             }
+            else if (!unpartitionedChildren.isEmpty()) {
+                if (!partitionedChildren.isEmpty()) {
+                    // add an gathering exchange above partitioned inputs and fold it into the set of unpartitioned inputs
+                    // NOTE: new symbols for ExchangeNode output are required in order to keep plan logically correct with new local union below
+
+                    List<Symbol> exchangeOutputLayout = node.getOutputSymbols().stream()
```

Comment:
Instead of special-casing it for this combination, why not keep a single branch like before and *always* allocate new names?